File: apps/macos/MonglePet/Runtime/PetSpeechRuntime.swift
import Foundation

nonisolated struct PetSpeechPresentation: Equatable, Sendable {
    let phraseID: UUID
    let text: String
    let displayDurationMilliseconds: Int64
    let theme: PetSpeechBubbleTheme
    let placement: PetSpeechBubblePlacementSettings

    init(
        phraseID: UUID,
        text: String,
        displayDurationMilliseconds: Int64,
        theme: PetSpeechBubbleTheme = .default,
        placement: PetSpeechBubblePlacementSettings = .default
    ) {
        self.phraseID = phraseID
        self.text = text
        self.displayDurationMilliseconds = displayDurationMilliseconds
        self.theme = theme
        self.placement = placement
    }
}

@MainActor
protocol PetSpeechScheduling: AnyObject {
    func schedule(after delay: Duration, action: @escaping () -> Void)
    func cancel()
}

@MainActor
final class RunLoopPetSpeechScheduler: NSObject, PetSpeechScheduling {
    private var timer: Timer?
    private var action: (() -> Void)?

    func schedule(after delay: Duration, action: @escaping () -> Void) {
        cancel()
        self.action = action
        let timer = Timer(
            timeInterval: max(delay.timeInterval, 0.001),
            target: self,
            selector: #selector(timerDidFire),
            userInfo: nil,
            repeats: false
        )
        RunLoop.main.add(timer, forMode: .common)
        self.timer = timer
    }

    func cancel() {
        timer?.invalidate()
        timer = nil
        action = nil
    }

    @objc
    private func timerDidFire() {
        timer = nil
        let pendingAction = action
        action = nil
        pendingAction?()
    }
}

@MainActor
final class PetSpeechRuntime {
    typealias PhrasePicker = ([PetSpeechPhrase], UUID?) -> PetSpeechPhrase?

    private enum PresentationSource: Equatable {
        case periodic
        case behavior(String)
    }

    private struct ActivePresentation {
        let presentation: PetSpeechPresentation
        let source: PresentationSource
        let displayMode: PetSpeechDisplayMode
    }

    private var settings: PetSpeechSettings = .default
    private let periodicScheduler: any PetSpeechScheduling
    private let dismissalScheduler: any PetSpeechScheduling
    private let phrasePicker: PhrasePicker
    private let onPresentationChange: (PetSpeechPresentation?) -> Void
    private var isAwake = false
    private var isSystemSuspended = false
    private var lastSequenceID: String?
    private var lastPeriodicPhraseID: UUID?
    private var lastBehaviorPhraseIDs: [String: UUID] = [:]
    private var activePresentation: ActivePresentation?
    private var isPeriodicScheduled = false
    private var isDismissalScheduled = false

    init(
        scheduler: any PetSpeechScheduling = RunLoopPetSpeechScheduler(),
        dismissalScheduler: any PetSpeechScheduling =
            RunLoopPetSpeechScheduler(),
        phrasePicker: @escaping PhrasePicker = { phrases, excludedID in
            let candidates: [PetSpeechPhrase]
            if phrases.count > 1, let excludedID {
                candidates = phrases.filter { $0.id != excludedID }
            } else {
                candidates = phrases
            }
            return candidates.randomElement()
        },
        onPresentationChange: @escaping (PetSpeechPresentation?) -> Void
    ) {
        periodicScheduler = scheduler
        self.dismissalScheduler = dismissalScheduler
        self.phrasePicker = phrasePicker
        self.onPresentationChange = onPresentationChange
    }

    func update(settings: PetSpeechSettings) {
        guard settings != self.settings else {
            return
        }
        self.settings = settings
        cancelPeriodic()
        cancelDismissal()
        lastSequenceID = nil
        lastPeriodicPhraseID = nil
        lastBehaviorPhraseIDs = [:]
        dismissCurrentPresentation()
        scheduleNextPeriodicPresentation()
    }

    func setAwake(_ isAwake: Bool) {
        guard isAwake != self.isAwake else {
            return
        }
        self.isAwake = isAwake
        if isAwake {
            scheduleNextPeriodicPresentation()
        } else {
            cancelPeriodic()
            cancelDismissal()
            dismissCurrentPresentation()
            lastSequenceID = nil
        }
    }

    func setSystemSuspended(_ isSuspended: Bool) {
        guard isSuspended != isSystemSuspended else {
            return
        }
        isSystemSuspended = isSuspended
        if isSuspended {
            cancelPeriodic()
            cancelDismissal()
            dismissCurrentPresentation()
            lastSequenceID = nil
        } else {
            scheduleNextPeriodicPresentation()
        }
    }

    func behaviorSequenceDidChange(_ sequenceID: String?) {
        guard sequenceID != lastSequenceID else {
            return
        }
        lastSequenceID = sequenceID
        guard canPresent else {
            return
        }

        if let sequenceID,
           let phrase = behaviorPhrase(for: sequenceID) {
            lastBehaviorPhraseIDs[sequenceID] = phrase.id
            present(phrase, source: .behavior(sequenceID))
            return
        }

        if settings.behaviorChangePolicy == .dismiss {
            dismissCurrentPresentation()
            scheduleNextPeriodicPresentation()
            return
        }

        guard let activePresentation else {
            ensurePeriodicPresentationScheduled()
            return
        }
        if case .behavior = activePresentation.source,
           activePresentation.displayMode == .untilNextPhrase {
            scheduleNextPeriodicPresentation()
        }
    }

    func stop() {
        cancelPeriodic()
        cancelDismissal()
        dismissCurrentPresentation()
        settings = .default
        isAwake = false
        isSystemSuspended = false
        lastSequenceID = nil
        lastPeriodicPhraseID = nil
        lastBehaviorPhraseIDs = [:]
    }

    func prepareForPetChange() {
        cancelPeriodic()
        cancelDismissal()
        dismissCurrentPresentation()
        settings = .default
        isAwake = false
        lastSequenceID = nil
        lastPeriodicPhraseID = nil
        lastBehaviorPhraseIDs = [:]
    }

    private var canPresent: Bool {
        settings.isEnabled && isAwake && !isSystemSuspended
    }

    private func scheduleNextPeriodicPresentation() {
        cancelPeriodic()
        guard
            canPresent,
            settings.periodicIsEnabled,
            !settings.periodicPhrases.isEmpty
        else {
            return
        }
        isPeriodicScheduled = true
        periodicScheduler.schedule(
            after: .milliseconds(settings.periodicIntervalMilliseconds)
        ) { [weak self] in
            self?.periodicTimerDidFire()
        }
    }

    private func periodicTimerDidFire() {
        isPeriodicScheduled = false
        guard canPresent else {
            cancelPeriodic()
            return
        }
        guard let phrase = nextPeriodicPhrase() else {
            return
        }
        lastPeriodicPhraseID = phrase.id
        present(phrase, source: .periodic)
    }

    private func present(
        _ phrase: PetSpeechPhrase,
        source: PresentationSource
    ) {
        cancelDismissal()
        if case .behavior = source {
            cancelPeriodic()
        }
        let presentation = PetSpeechPresentation(
            phraseID: phrase.id,
            text: phrase.text,
            displayDurationMilliseconds:
                phrase.displayDurationMilliseconds,
            theme: settings.theme,
            placement: settings.placement
        )
        activePresentation = ActivePresentation(
            presentation: presentation,
            source: source,
            displayMode: phrase.displayMode
        )
        onPresentationChange(presentation)

        switch phrase.displayMode {
        case .timed:
            scheduleDismissal(
                phraseID: phrase.id,
                after: phrase.displayDurationMilliseconds
            )
        case .untilNextPhrase:
            scheduleNextPeriodicPresentation()
        }
    }

    private func scheduleDismissal(
        phraseID: UUID,
        after milliseconds: Int64
    ) {
        cancelDismissal()
        isDismissalScheduled = true
        dismissalScheduler.schedule(
            after: .milliseconds(milliseconds)
        ) { [weak self] in
            self?.dismissalTimerDidFire(for: phraseID)
        }
    }

    private func dismissalTimerDidFire(for phraseID: UUID) {
        isDismissalScheduled = false
        guard activePresentation?.presentation.phraseID == phraseID else {
            return
        }
        dismissCurrentPresentation()
        scheduleNextPeriodicPresentation()
    }

    private func dismissCurrentPresentation() {
        cancelDismissal()
        guard activePresentation != nil else {
            return
        }
        activePresentation = nil
        onPresentationChange(nil)
    }

    private func ensurePeriodicPresentationScheduled() {
        guard !isPeriodicScheduled else {
            return
        }
        scheduleNextPeriodicPresentation()
    }

    private func cancelPeriodic() {
        periodicScheduler.cancel()
        isPeriodicScheduled = false
    }

    private func cancelDismissal() {
        dismissalScheduler.cancel()
        isDismissalScheduled = false
    }

    private func behaviorPhrase(
        for sequenceID: String
    ) -> PetSpeechPhrase? {
        phrasePicker(
            settings.behaviorPhrases.filter {
                $0.trigger == .sequence(sequenceID)
            },
            lastBehaviorPhraseIDs[sequenceID]
        )
    }

    private func nextPeriodicPhrase() -> PetSpeechPhrase? {
        let phrases = settings.periodicPhrases
        switch settings.periodicOrder {
        case .random:
            return phrasePicker(phrases, lastPeriodicPhraseID)
        case .sequential:
            guard
                let lastPeriodicPhraseID,
                let index = phrases.firstIndex(
                    where: { $0.id == lastPeriodicPhraseID }
                )
            else {
                return phrases.first
            }
            let nextIndex = phrases.index(after: index)
            return nextIndex == phrases.endIndex
                ? phrases.first
                : phrases[nextIndex]
        }
    }

}

private extension Duration {
    var timeInterval: TimeInterval {
        let components = self.components
        return TimeInterval(components.seconds)
            + TimeInterval(components.attoseconds)
                / 1_000_000_000_000_000_000
    }
}

File: apps/windows/src/MonglePet.Settings/PetSpeechRuntime.cs
namespace MonglePet.Settings;

public sealed record PetSpeechPresentation(
    Guid PhraseId,
    string Text,
    long DisplayDurationMilliseconds,
    PetSpeechBubbleTheme Theme,
    PetSpeechBubblePlacementSettings Placement);

public interface IPetSpeechScheduler
{
    void Schedule(TimeSpan delay, Action action);

    void Cancel();
}

public sealed class PetSpeechRuntime : IDisposable
{
    public delegate PetSpeechPhrase? PhrasePicker(
        IReadOnlyList<PetSpeechPhrase> phrases,
        Guid? excludedId);

    private enum PresentationSourceKind
    {
        Periodic,
        Behavior,
    }

    private sealed record ActivePresentation(
        PetSpeechPresentation Presentation,
        PresentationSourceKind Source,
        string? SequenceId,
        PetSpeechDisplayMode DisplayMode);

    private PetSpeechSettings _settings = PetSpeechSettings.Default;
    private readonly IPetSpeechScheduler _periodicScheduler;
    private readonly IPetSpeechScheduler _dismissalScheduler;
    private readonly PhrasePicker _phrasePicker;
    private readonly Action<PetSpeechPresentation?> _onPresentationChange;
    private readonly Dictionary<string, Guid> _lastBehaviorPhraseIds =
        new(StringComparer.Ordinal);
    private bool _isAwake;
    private bool _isSystemSuspended;
    private string? _lastSequenceId;
    private Guid? _lastPeriodicPhraseId;
    private ActivePresentation? _activePresentation;
    private bool _isPeriodicScheduled;
    private bool _disposed;

    public PetSpeechRuntime(
        IPetSpeechScheduler periodicScheduler,
        IPetSpeechScheduler dismissalScheduler,
        Action<PetSpeechPresentation?> onPresentationChange,
        PhrasePicker? phrasePicker = null)
    {
        ArgumentNullException.ThrowIfNull(periodicScheduler);
        ArgumentNullException.ThrowIfNull(dismissalScheduler);
        ArgumentNullException.ThrowIfNull(onPresentationChange);
        _periodicScheduler = periodicScheduler;
        _dismissalScheduler = dismissalScheduler;
        _onPresentationChange = onPresentationChange;
        _phrasePicker = phrasePicker ?? PickRandomPhrase;
    }

    public string Status
    {
        get
        {
            int periodicPhraseCount = PeriodicPhrases().Count;
            if (!_settings.IsEnabled)
            {
                return "말풍선 사용 안 함";
            }
            if (!_isAwake)
            {
                return "펫이 자는 중";
            }
            if (_isSystemSuspended)
            {
                return "잠금 또는 절전으로 일시 정지";
            }
            if (!_settings.PeriodicIsEnabled)
            {
                return "주기 대사 사용 안 함";
            }
            if (periodicPhraseCount == 0)
            {
                return "등록된 주기 대사 없음";
            }
            if (_activePresentation is { } active)
            {
                return $"대사 표시 중 · {active.Presentation.Text}";
            }
            return _isPeriodicScheduled
                ? $"다음 주기 대사 예약됨 · {_settings.PeriodicIntervalMilliseconds / 1_000d:0.#}초"
                : "주기 대사 예약 대기 중";
        }
    }

    public void Update(PetSpeechSettings settings)
    {
        ArgumentNullException.ThrowIfNull(settings);
        ThrowIfDisposed();
        if (settings == _settings)
        {
            return;
        }

        _settings = settings;
        CancelPeriodic();
        CancelDismissal();
        _lastSequenceId = null;
        _lastPeriodicPhraseId = null;
        _lastBehaviorPhraseIds.Clear();
        DismissCurrentPresentation();
        ScheduleNextPeriodicPresentation();
    }

    public void SetAwake(bool isAwake)
    {
        ThrowIfDisposed();
        if (isAwake == _isAwake)
        {
            return;
        }

        _isAwake = isAwake;
        if (isAwake)
        {
            ScheduleNextPeriodicPresentation();
        }
        else
        {
            CancelPeriodic();
            CancelDismissal();
            DismissCurrentPresentation();
            _lastSequenceId = null;
        }
    }

    public void SetSystemSuspended(bool isSuspended)
    {
        ThrowIfDisposed();
        if (isSuspended == _isSystemSuspended)
        {
            return;
        }

        _isSystemSuspended = isSuspended;
        if (isSuspended)
        {
            CancelPeriodic();
            CancelDismissal();
            DismissCurrentPresentation();
            _lastSequenceId = null;
        }
        else
        {
            ScheduleNextPeriodicPresentation();
        }
    }

    public void BehaviorSequenceDidChange(string? sequenceId)
    {
        ThrowIfDisposed();
        if (string.Equals(sequenceId, _lastSequenceId, StringComparison.Ordinal))
        {
            return;
        }

        _lastSequenceId = sequenceId;
        if (!CanPresent)
        {
            return;
        }

        if (sequenceId is not null && BehaviorPhrase(sequenceId) is { } phrase)
        {
            _lastBehaviorPhraseIds[sequenceId] = phrase.Id;
            Present(
                phrase,
                PresentationSourceKind.Behavior,
                sequenceId);
            return;
        }

        if (_settings.BehaviorChangePolicy == PetSpeechBehaviorChangePolicy.Dismiss)
        {
            DismissCurrentPresentation();
            ScheduleNextPeriodicPresentation();
            return;
        }

        if (_activePresentation is null)
        {
            EnsurePeriodicPresentationScheduled();
            return;
        }

        if (_activePresentation.Source == PresentationSourceKind.Behavior &&
            _activePresentation.DisplayMode == PetSpeechDisplayMode.UntilNextPhrase)
        {
            ScheduleNextPeriodicPresentation();
        }
    }

    public void PrepareForPetChange()
    {
        ThrowIfDisposed();
        Reset();
    }

    public void Stop()
    {
        ThrowIfDisposed();
        Reset();
        _isSystemSuspended = false;
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        Reset();
        _disposed = true;
        if (_periodicScheduler is IDisposable periodicDisposable)
        {
            periodicDisposable.Dispose();
        }
        if (!ReferenceEquals(_periodicScheduler, _dismissalScheduler) &&
            _dismissalScheduler is IDisposable dismissalDisposable)
        {
            dismissalDisposable.Dispose();
        }
        GC.SuppressFinalize(this);
    }

    private bool CanPresent =>
        _settings.IsEnabled && _isAwake && !_isSystemSuspended;

    private void ScheduleNextPeriodicPresentation()
    {
        CancelPeriodic();
        if (!CanPresent ||
            !_settings.PeriodicIsEnabled ||
            !PeriodicPhrases().Any())
        {
            return;
        }

        _isPeriodicScheduled = true;
        _periodicScheduler.Schedule(
            TimeSpan.FromMilliseconds(_settings.PeriodicIntervalMilliseconds),
            PeriodicTimerDidFire);
    }

    private void PeriodicTimerDidFire()
    {
        _isPeriodicScheduled = false;
        if (!CanPresent)
        {
            CancelPeriodic();
            return;
        }

        PetSpeechPhrase? phrase = NextPeriodicPhrase();
        if (phrase is null)
        {
            return;
        }

        _lastPeriodicPhraseId = phrase.Id;
        Present(phrase, PresentationSourceKind.Periodic, null);
    }

    private void Present(
        PetSpeechPhrase phrase,
        PresentationSourceKind source,
        string? sequenceId)
    {
        CancelDismissal();
        if (source == PresentationSourceKind.Behavior)
        {
            CancelPeriodic();
        }

        var presentation = new PetSpeechPresentation(
            phrase.Id,
            phrase.Text,
            phrase.DisplayDurationMilliseconds,
            _settings.Theme,
            _settings.Placement);
        _activePresentation = new ActivePresentation(
            presentation,
            source,
            sequenceId,
            phrase.DisplayMode);
        _onPresentationChange(presentation);

        if (phrase.DisplayMode == PetSpeechDisplayMode.Timed)
        {
            _dismissalScheduler.Schedule(
                TimeSpan.FromMilliseconds(phrase.DisplayDurationMilliseconds),
                () => DismissalTimerDidFire(phrase.Id));
        }
        else
        {
            ScheduleNextPeriodicPresentation();
        }
    }

    private void DismissalTimerDidFire(Guid phraseId)
    {
        if (_activePresentation?.Presentation.PhraseId != phraseId)
        {
            return;
        }

        DismissCurrentPresentation();
        ScheduleNextPeriodicPresentation();
    }

    private void DismissCurrentPresentation()
    {
        CancelDismissal();
        if (_activePresentation is null)
        {
            return;
        }

        _activePresentation = null;
        _onPresentationChange(null);
    }

    private void EnsurePeriodicPresentationScheduled()
    {
        if (!_isPeriodicScheduled)
        {
            ScheduleNextPeriodicPresentation();
        }
    }

    private void CancelPeriodic()
    {
        _periodicScheduler.Cancel();
        _isPeriodicScheduled = false;
    }

    private void CancelDismissal() => _dismissalScheduler.Cancel();

    private PetSpeechPhrase? BehaviorPhrase(string sequenceId)
    {
        IReadOnlyList<PetSpeechPhrase> phrases = _settings.Phrases
            .Where(phrase => phrase.Trigger is PetSpeechTrigger.Sequence trigger &&
                string.Equals(trigger.SequenceId, sequenceId, StringComparison.Ordinal))
            .ToList();
        if (phrases.Count == 0)
        {
            return null;
        }
        _lastBehaviorPhraseIds.TryGetValue(sequenceId, out Guid lastId);
        return _phrasePicker(phrases, lastId == Guid.Empty ? null : lastId);
    }

    private PetSpeechPhrase? NextPeriodicPhrase()
    {
        IReadOnlyList<PetSpeechPhrase> phrases = PeriodicPhrases();
        if (_settings.PeriodicOrder == PetSpeechPeriodicOrder.Random)
        {
            return _phrasePicker(phrases, _lastPeriodicPhraseId);
        }

        if (_lastPeriodicPhraseId is not Guid lastId)
        {
            return phrases.FirstOrDefault();
        }

        int index = phrases.ToList().FindIndex(phrase => phrase.Id == lastId);
        return index < 0 || index + 1 >= phrases.Count
            ? phrases.FirstOrDefault()
            : phrases[index + 1];
    }

    private IReadOnlyList<PetSpeechPhrase> PeriodicPhrases() =>
        _settings.Phrases
            .Where(phrase => phrase.Trigger is PetSpeechTrigger.Periodic)
            .ToList();

    private void Reset()
    {
        CancelPeriodic();
        CancelDismissal();
        DismissCurrentPresentation();
        _settings = PetSpeechSettings.Default;
        _isAwake = false;
        _lastSequenceId = null;
        _lastPeriodicPhraseId = null;
        _lastBehaviorPhraseIds.Clear();
    }

    private static PetSpeechPhrase? PickRandomPhrase(
        IReadOnlyList<PetSpeechPhrase> phrases,
        Guid? excludedId)
    {
        IReadOnlyList<PetSpeechPhrase> candidates =
            phrases.Count > 1 && excludedId is not null
                ? phrases.Where(phrase => phrase.Id != excludedId).ToList()
                : phrases;
        return candidates.Count == 0
            ? null
            : candidates[Random.Shared.Next(candidates.Count)];
    }

    private void ThrowIfDisposed() => ObjectDisposedException.ThrowIf(_disposed, this);
}
